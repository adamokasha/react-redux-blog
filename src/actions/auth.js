import axios from "axios";

export const login = (displayName, id, role) => ({
  type: "LOGIN",
  displayName,
  id,
  role
});

export const startSignup = (displayName, email, password) => {
  return async dispatch => {
    try {
      const response = await axios.post(`${process.env.API_URI}/users`, {
        displayName,
        email,
        password
      });

      const user = {
        displayName: response.data.displayName,
        id: response.data._id,
        token: response.headers["x-auth"],
        role: response.data.role
      };
      localStorage.setItem("auth", JSON.stringify(user));

      dispatch(
        login(response.data.displayName, response.data._id, response.data.role)
      );
    } catch (e) {
      throw e.response.data.errors;
    }
  };
};

export const startLogin = (email, password) => {
  return async dispatch => {
    try {
      const response = await axios.post(`${process.env.API_URI}/users/login`, {
        email,
        password
      });

      const user = {
        displayName: response.data.displayName,
        id: response.data._id,
        token: response.headers["x-auth"],
        role: response.data.role
      };
      localStorage.setItem("auth", JSON.stringify(user));

      dispatch(
        login(response.data.displayName, response.data._id, response.data.role)
      );
    } catch (e) {
      throw new Error(e.response.data.error);
    }
  };
};

export const startLogout = () => {
  return async dispatch => {
    try {
      const { token } = JSON.parse(localStorage.getItem("auth"));
      localStorage.removeItem("auth");
      await axios.delete(`${process.env.API_URI}/users/me/token`, {
        headers: {
          "x-auth": token
        }
      });
      dispatch(logout());
    } catch (e) {
      localStorage.removeItem("auth");
      dispatch(logout());
    }
  };
};

export const logout = () => ({
  type: "LOGOUT"
});
