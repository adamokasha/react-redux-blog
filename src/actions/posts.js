import axios from "axios";

export const setPosts = posts => ({
  type: "SET_POSTS",
  posts
});

export const startSetPosts = () => {
  return async dispatch => {
    try {
      const response = await axios.get(
        `${process.env.REACT_APP_API_URL}/posts`
      );
      dispatch(setPosts(response.data.posts));
    } catch (e) {
      throw new Error("Could not retrieve posts...");
    }
  };
};

export const addPost = post => ({
  type: "ADD_POST",
  post
});

export const startAddPost = post => {
  return async dispatch => {
    const { token } = JSON.parse(localStorage.getItem("auth"));
    try {
      const response = await axios({
        url: `${process.env.REACT_APP_API_URL}/posts`,
        method: "post",
        headers: {
          "x-auth": token
        },
        data: post
      });

      dispatch(addPost(response.data.post));
    } catch (e) {
      throw new Error(e.response.data.error);
    }
  };
};

export const editPost = updates => ({
  type: "EDIT_POST",
  updates
});

export const startEditPost = (post, id) => {
  return async dispatch => {
    const { token } = JSON.parse(localStorage.getItem("auth"));
    try {
      const response = await axios({
        url: `${process.env.REACT_APP_API_URL}/posts/${id}`,
        method: "patch",
        headers: {
          "x-auth": token
        },
        data: post
      });

      dispatch(editPost(response.data.post));
    } catch (e) {
      throw new Error(e.response.data.error);
    }
  };
};

export const deletePost = id => ({
  type: "DELETE_POST",
  id
});

export const startDeletePost = id => {
  return async dispatch => {
    const { token } = JSON.parse(localStorage.getItem("auth"));
    try {
      const response = await axios({
        url: `${process.env.REACT_APP_API_URL}/posts/${id}/`,
        method: "delete",
        headers: {
          "x-auth": token
        }
      });

      dispatch(deletePost(id));
    } catch (e) {
      throw new Error(e.response.data.error);
    }
  };
};

export const addComment = (id, comment) => ({
  type: "ADD_COMMENT",
  id,
  comment
});

export const startAddComment = (id, comment) => {
  return async dispatch => {
    try {
      const { token } = JSON.parse(localStorage.getItem("auth"));
      const response = await axios({
        url: `${process.env.REACT_APP_API_URL}/posts/${id}/comments`,
        method: "post",
        headers: {
          "x-auth": token
        },
        data: {
          comment
        }
      });

      dispatch(addComment(id, response.data));
    } catch (e) {
      throw new Error(e.response.data.error);
    }
  };
};
