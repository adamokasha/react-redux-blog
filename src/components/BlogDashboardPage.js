import React from "react";
import { connect } from "react-redux";

import Header from "./Header";
import PostsList from "./PostsList";

export const BlogDashboardPage = ({ displayName }) => {
  function renderWelcome() {
    if (displayName) {
      return <h3>Welcome {displayName}.</h3>;
    }
    return;
  }
  return (
    <div>
      <Header />
      <div id="landing">
        <div className="primary-overlay">
          <div className="container h-100">
            <div className="row hero-text">
              <div className="col-lg-7">
                <h1 className="text-white mb-5">
                  React Blog - Responsive blog styled with Bootstrap 4
                </h1>
                <a
                  href="https://github.com/adamokasha/react-redux-blog"
                  target="_blank"
                  className="text-uppercase text-white hero-link"
                >
                  <u>Learn More</u>
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <br />
      <div className="container mb-3">{renderWelcome()}</div>
      <div>
        <PostsList />
      </div>
    </div>
  );
};

const mapStateToProps = state => ({
  displayName: state.auth.displayName
});

export default connect(mapStateToProps)(BlogDashboardPage);
