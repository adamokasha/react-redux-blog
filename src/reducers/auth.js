const defaultAuthState = () => {
  if (localStorage.getItem("auth")) {
    return JSON.parse(localStorage.getItem("auth"));
  } else {
    return {};
  }
};

const authReducer = (state = defaultAuthState(), action) => {
  switch (action.type) {
    case "LOGIN":
      return {
        displayName: action.displayName,
        id: action.id,
        role: action.role
      };
    case "LOGOUT":
      return {};
    default:
      return state;
  }
};

export default authReducer;
