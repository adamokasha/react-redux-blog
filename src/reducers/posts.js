const postsReducer = (state = [], action) => {
  switch (action.type) {
    case "SET_POSTS":
      return action.posts;
    case "ADD_POST":
      return [...state, action.post];
    case "EDIT_POST":
      return state.map(post => {
        if (post._id === action.updates._id) {
          return { ...post, ...action.updates };
        } else {
          return post;
        }
      });
    case "ADD_COMMENT":
      return state.map(post => {
        if (post._id === action.id) {
          return {
            ...post,
            comments: [...post.comments, action.comment]
          };
        } else {
          return post;
        }
      });
    case "DELETE_POST":
      return state.filter(post => {
        if (post._id !== action.id) {
          return post;
        }
      });
    default:
      return state;
  }
};

export default postsReducer;
