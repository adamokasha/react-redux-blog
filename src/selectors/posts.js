export const sortPostsByDate = postsArray => {
  const newPostsArray = [].concat(postsArray);
  return newPostsArray.sort((a, b) => {
    if (a.createdAt > b.createdAt) {
      return -1;
    }
    if (a.createdAt < b.createdAt) {
      return 1;
    }
  });
};

export const sortCommentsByDate = commentsArray => {
  const newCommentsArray = [].concat(commentsArray);
  return newCommentsArray.sort((a, b) => {
    if (a.createdAt > b.createdAt) {
      return -1;
    }
    if (a.createdAt < b.createdAt) {
      return 1;
    }
  });
};
