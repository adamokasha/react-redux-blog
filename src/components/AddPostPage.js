import React from "react";
import { connect } from "react-redux";

import PostForm from "./PostForm";
import { startAddPost } from "../actions/posts";

export class AddPostPage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      error: ""
    };
  }
  componentDidMount() {
    window.scrollTo(0, 0);
  }
  renderError = () => {
    if (this.state.error) {
      return (
        <div className="error-msg alert alert-danger" role="alert">
          <p>{this.state.error}</p>
        </div>
      );
    }
  };
  onSubmit = async post => {
    try {
      await this.props.startAddPost(post);
      this.props.history.push("/");
    } catch (e) {
      this.setState({ error: e.message });
    }
  };
  render() {
    return (
      <div className="container">
        <h3 className="text-center mt-4">Add Post</h3>
        {this.renderError()}
        <PostForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddPost: post => dispatch(startAddPost(post))
});

export default connect(
  null,
  mapDispatchToProps
)(AddPostPage);
