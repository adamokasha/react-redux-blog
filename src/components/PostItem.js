import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import moment from "moment";

import Header from "./Header";
import CommentBox from "./CommentBox";
import { startSetPosts } from "../actions/posts";

export class PostItem extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      error: ""
    };
  }
  async componentDidMount() {
    window.scrollTo(0, 0);
    if (!this.props.post) {
      try {
        await this.props.startSetPosts();
      } catch (e) {
        this.setState({ error: e.message });
      }
    }
  }
  renderEditButton() {
    if (this.props.auth.role === "admin") {
      return (
        <Link
          to={`/posts/${this.props.post._id}/edit`}
          className="btn btn-dark btn-md float-right"
        >
          <i className="far fa-edit" /> Edit
        </Link>
      );
    }
  }
  renderMainImage(post) {
    if (post.mainImage) {
      return (
        <img
          src={post.mainImage}
          className="img-fluid rounded float-left mr-3 mb-3 d-none d-md-inline-block w-50"
        />
      );
    }
  }
  renderPost() {
    if (this.props.post) {
      return (
        <div className="container post-item">
          <div>
            {this.renderEditButton()}
            <h2>{this.props.post.title}</h2>
            <p className="text-muted">
              Posted by {this.props.post.author} on{" "}
              {moment(this.props.post.createdAt).format("MMMM Do YYYY")}.
            </p>
            {this.renderMainImage(this.props.post)}
            <p>{this.props.post.body}</p>
          </div>
          <CommentBox
            id={this.props.post._id}
            comments={this.props.post.comments}
          />
        </div>
      );
    }
    return (
      <div className="container post-item">
        <div className="alert alert-danger" role="alert">
          Could not retrieve post...
          <Link to="/" className="alert-link">
            Click here
          </Link>{" "}
          to go back to dashboard.
        </div>
      </div>
    );
  }
  render() {
    return (
      <div>
        <Header />
        {this.renderPost()}
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  auth: state.auth,
  post: state.posts.find(post => post._id === props.match.params.id) || null
});

export default connect(
  mapStateToProps,
  { startSetPosts }
)(PostItem);
