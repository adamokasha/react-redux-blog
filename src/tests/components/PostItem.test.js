import React from "react";
import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import { PostItem } from "../../components/PostItem";
import posts from "../fixtures/posts";
import { withAdminRole, withUserRole } from "../fixtures/auth";

configure({ adapter: new Adapter() });

test("should render a single post", () => {
  const wrapper = shallow(<PostItem post={posts[0]} auth={withUserRole} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render edit button if user role is admin", () => {
  const wrapper = shallow(<PostItem post={posts[0]} auth={withAdminRole} />);
  expect(wrapper.find("Link i.fa-edit").length).toBe(1);
});

test("should not render edit button if user role is user", () => {
  const wrapper = shallow(<PostItem post={posts[0]} auth={withUserRole} />);
  expect(wrapper.find("Link button").length).toBe(0);
});
