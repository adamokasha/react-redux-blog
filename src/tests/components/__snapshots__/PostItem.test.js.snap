// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a single post 1`] = `
<div>
  <Connect(Header) />
  <div
    className="container post-item"
  >
    <div>
      <h2>
        The first post
      </h2>
      <p
        className="text-muted"
      >
        Posted by 
        UserOne
         on
         
        August 2nd 2018
        .
      </p>
      <img
        className="img-fluid rounded float-left mr-3 mb-3 d-none d-md-inline-block w-50"
        src="http://mysite.com/img.jpg"
      />
      <p>
        This is the first post. The body has at least 24 characters.
      </p>
    </div>
    <Connect(CommentBox)
      comments={
        Array [
          Object {
            "_id": "5b62cdeb62105f402c9a8749",
            "comment": "A Comment",
            "createdAt": 1533201893560,
            "createdBy": "theadmin",
          },
          Object {
            "_id": "5b62cde562105f402c9a8748",
            "comment": "Another Comment",
            "createdAt": 1533201899157,
            "createdBy": "theadmin",
          },
        ]
      }
      id="5b62cb86efcae617d8a0bef9"
    />
  </div>
</div>
`;
